src/accelerate/test_utils/scripts/test_distributed_data_loop.py

```diff
@@ -260,6 +265,67 @@ def test_data_loader(data_loader, accelerator):
     ), "Not all the dataset elements have been iterated in an epoch due to duplication of samples across processes."
 
 
+def test_stateful_dataloader(accelerator):
+    old_dataloader_config = accelerator.dataloader_config
+    accelerator.dataloader_config = DataLoaderConfiguration(use_stateful_dataloader=True)
+    prepared_dl = create_dataloader(
+        accelerator, dataset_size=32 * accelerator.num_processes, batch_size=4, iterable=True, shuffle=True
+    )
+    untrained_batches = []
+    # Calculate what step that will be
+    total_batches = 32 * accelerator.num_processes // (4 * accelerator.num_processes)
+    last_batch_num = total_batches - 1
+    for step, batch in enumerate(prepared_dl):
+        # Step just before
+        if step == last_batch_num - 1:
+            state_dict = prepared_dl.state_dict()
+        if step >= last_batch_num:
+            # Otherwise grab the "unseen" batches
+            untrained_batches.append(batch)
+    not_skipped_batches = accelerator.gather(untrained_batches)
+    prepared_dl.load_state_dict(state_dict)
+    resumed_batches = []
+    for batch in prepared_dl:
+        resumed_batches.append(batch)
+    resumed_batches = accelerator.gather(resumed_batches)
+    for b1, b2 in zip(not_skipped_batches, resumed_batches):
+        for v1, v2 in zip(b1, b2):
+            assert torch.equal(v1, v2), f"Batch {b1} and {b2} are not equal"
+
+    accelerator.dataloader_config = old_dataloader_config
```

Comment:
Does this need to be inside a `try ... finally` if this test fails but others have yet to run?